### PR TITLE
Implement assignment groups and graceful failure

### PR DESCRIPTION
**Pull request by Laremere**
Resolves https://github.com/googleforgames/open-match/issues/1155

Rewrote the tests to mostly be not parameterized, since they are testing different scenarios, I find straightforward tests to be simple and easier to understand.

### CODE REVIEW

File: internal/statestore/public.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ type Service interface {
 	GetTickets(ctx context.Context, ids []string) ([]*pb.Ticket, error)
 
 	// UpdateAssignments update the match assignments for the input ticket ids
-	UpdateAssignments(ctx context.Context, ids []string, assignment *pb.Assignment) error
+	UpdateAssignments(ctx context.Context, req *pb.AssignTicketsRequest) (*pb.AssignTicketsResponse, error)
```
Comment (yfei1) [inline internal/statestore/public.go]: can you update the outdated comments?

File: internal/statestore/redis.go [hunk 2]
```diff
@@ -476,78 +476,83 @@ func (rb *redisBackend) GetTickets(ctx context.Context, ids []string) ([]*pb.Tic
...
+				TicketId: ids[i],
+				Cause:    pb.AssignmentFailure_TICKET_NOT_FOUND,
+			})
+		} else {
+			t := &pb.Ticket{}
+			err = proto.Unmarshal(ticketByte, t)
+			if err != nil {
+				redisLogger.WithFields(logrus.Fields{
+					"key": ids[i],
+				}).WithError(err).Error("Failed to unmarshal ticket from redis.")
```
Comment (yfei1) [inline internal/statestore/redis.go]: lowercase

File: internal/statestore/redis.go [hunk 3]
```diff
@@ -476,78 +476,83 @@ func (rb *redisBackend) GetTickets(ctx context.Context, ids []string) ([]*pb.Tic
...
-		}
-		t := &pb.Ticket{}
-		err = proto.Unmarshal(ticketByte, t)
-		if err != nil {
-			redisLogger.WithFields(logrus.Fields{
-				"key": ids[i],
-			}).WithError(err).Error("Failed to unmarshal ticket from redis.")
-			return status.Errorf(codes.Internal, "%v", err)
+			resp.Failed = append(resp.Failed, &pb.AssignmentFailure{
+				TicketId: ids[i],
```
Comment (yfei1) [inline internal/statestore/redis.go]: `idsI[i]` works at here. do we really need the `ids` slice?

Comment (Laremere) [reply internal/statestore/redis.go]: It's also used to create an AssignmentFailure, which needs a string.

Comment (yfei1) [reply internal/statestore/redis.go]: is type assertion not working?

Comment (Laremere) [reply internal/statestore/redis.go]: I'd rather not use a type assertion if I don't have to.

**Pull request merged**